modular_skyrat/master_files/code/modules/mob/living/sillicon/robot.dm

```diff
@@ -26,46 +26,97 @@
 
 /datum/component/robot_smoke/RegisterWithParent()
 	add_verb(parent, /mob/living/silicon/robot/proc/toggle_smoke)
+	RegisterSignal(parent, COMSIG_ATOM_DIR_CHANGE, .proc/dir_change)
 
 /datum/component/robot_smoke/UnregisterFromParent()
 	remove_verb(parent, /mob/living/silicon/robot/proc/toggle_smoke)
+	UnregisterSignal(parent, COMSIG_ATOM_DIR_CHANGE)
 
 /datum/component/robot_smoke/Destroy()
 	return ..()
 
+/datum/component/robot_smoke/proc/dir_change(datum/source, olddir, newdir)
+	SIGNAL_HANDLER
+
+	var/atom/movable/movable_parent = parent
+
+	if(!movable_parent.particles)
+		return
+
+	var/truedir = movable_parent.dir
+	if(newdir && truedir != newdir)
+		truedir = newdir
+
+	switch(truedir)
+		if(NORTH)
+			movable_parent.particles.position = list(-6, 12, 0)
+			movable_parent.particles.drift = generator("vector", list(0, 0.4), list(0.2, -0.2))
+		if(EAST)
+			movable_parent.particles.position = list(-6, 12, 0)
+			movable_parent.particles.drift = generator("vector", list(0, 0.4), list(-0.8, 0.2))
+		if(SOUTH)
+			movable_parent.particles.position = list(5, 12, 0)
+			movable_parent.particles.drift = generator("vector", list(0, 0.4), list(0.2, -0.2))
+		if(WEST)
+			movable_parent.particles.position = list(6, 12, 0)
+			movable_parent.particles.drift = generator("vector", list(0, 0.4), list(0.8, -0.2))
+
+
 /mob/living/silicon/robot/proc/toggle_smoke()
 	set name = "Toggle smoke"
 	set category = "AI Commands"
 
 	if(particles)
-		QDEL_NULL(particles)
+		smoke_disappear()
 	else if (!stat && !robot_resting)
+		do_jitter_animation(10)
+		playsound(src, 'modular_skyrat/master_files/sound/effects/robot_smoke.ogg', 50)
 		particles = new /particles/smoke/robot()
 
+/mob/living/silicon/robot/proc/smoke_disappear()
```

Comment:
Why can't this be part of disposing()

Reply:
>disposing()

What is this? I can't find anything about this in our code except for a COMSIG about being shoved into disposals.

Reply:
This datum only exists on Goonstation